### PR TITLE
Add date_of_death_reported_at to track when veteran date_of_death changes

### PR DESCRIPTION
**Pull request by bgantick**
Resolves [CASEFLOW-46](https://vajira.max.gov/browse/CASEFLOW-46)

### Description
Method/migration to add `death_reported_at` to `Veteran` model

### Acceptance Criteria
- [x] Code compiles correctly

#### Update an existing veteran:
1. In rails console, pull up a veteran that doesn't currently have a `date_of_death`:
```
v = Veteran.where(:file_number => 152003980).first
```
2. Update `date_of_death` and save:
```
v.date_of_death = Time.utc(2020, 3, 4, 12, 0, 0)
v.save
```
3. Ensure `date_of_death_reported_at` is updated accordingly (`v` should now have a `date_of_death_reported_at` with a timestamp that is equal to the save time)

### Database Changes

* [x] Column comments updated
* [x] Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`)
* [x] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [x] Query profiling performed (eyeball Rails log, check bullet and fasterer output)
* [x] DB schema docs updated with `make docs` (after running `make migrate`)
* [x] #appeals-schema notified with summary and link to this PR

### CODE REVIEW

**Comment by meckila-navapbc** [comment]
This LGTM so far

File: db/migrate/20210126194054_add_death_reported_on_date_to_veteran.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class AddDeathReportedOnDateToVeteran < Caseflow::Migration
+  def change
+    add_column :veterans,
+               :death_reported_on_date,
+               :datetime,
```
Comment (nanotone) [inline db/migrate/20210126194054_add_death_reported_on_date_to_veteran.rb:5]: Following the principle of least surprise, I think ActiveRecord and by extension Caseflow have a loose convention of datetimes (representing events happening within the app) ending in `_at`. e.g. `created_at`, `updated_at`, `established_at`, `contention_removed_at`, `decision_sync_attempted_at`.

So the nitpick about the name here is that both "on" and "date" suggest date-precision, whereas you're actually storing datetime precision.

**Review by nanotone — COMMENTED** [review]
This is really nice and elegant! I have one nitpick and one question/concern.

File: app/models/veteran.rb [hunk 2]
```diff
@@ -37,6 +37,8 @@ class Veteran < CaseflowRecord
 
   delegate :full_address, to: :address
 
+  before_save :set_death_reported_on_date!
```
Comment (yoomlam) [inline app/models/veteran.rb:40]: IMO, it's easier to determine relevance of this hook if you check `if date_of_death_changed?` here.
```suggestion
  before_save :set_death_reported_on_date! if date_of_death_changed?
```

File: db/migrate/20210126194054_add_death_reported_on_date_to_veteran.rb (hunk 1 above)
Comment (bgantick) [reply db/migrate/20210126194054_add_death_reported_on_date_to_veteran.rb]: Good call. Changed to `death_reported_at`

File: app/models/veteran.rb [hunk 3]
```diff
@@ -37,6 +37,8 @@ class Veteran < CaseflowRecord
 
   delegate :full_address, to: :address
 
+  before_save :set_death_reported_at!
```
Comment (yoomlam) [inline app/models/veteran.rb:40]: IMO, it's easier to determine relevance of this hook if you check `if date_of_death_changed?` here.
```suggestion
  before_save : set_death_reported_at! if :date_of_death_changed?
```
or maybe `if :will_save_change_to_date_of_death?` https://stackoverflow.com/questions/24314584/run-a-callback-only-if-an-attribute-has-changed-in-rails

File: spec/models/veteran_spec.rb [hunk 4]
```diff
@@ -848,6 +848,10 @@
       it "is true" do
         is_expected.to eq(true)
       end
+
+      it "updates death_reported_at" do
+        expect(veteran.death_reported_at).to eq(Time.zone.now)
```
Comment (yoomlam) [inline spec/models/veteran_spec.rb:853]: Is this a potential flake if the time difference is a few milliseconds?

Comment (bgantick) [reply spec/models/veteran_spec.rb:853]: Really good question. I believe that's why we're using Timecop in `before do` at the top of the file? Maybe I'm not fully understanding how Timecop works.

Comment (yoomlam) [reply spec/models/veteran_spec.rb:853]: I see there's a `Timecop.freeze` on line 15, which should freeze time, but you can insert a `binding.pry` or sleep to make sure the time is indeed frozen.

File: app/models/veteran.rb (hunk 3 above)
Comment (bgantick) [reply app/models/veteran.rb]: Good optimization! Updated.

**Comment by nanotone** [comment]
Answered my own question:

```ruby
> v = Veteran.new
> v.will_save_change_to_date_of_death?
=> false
> v.date_of_death = Time.zone.yesterday
=> Sun, 31 Jan 2021
> v.will_save_change_to_date_of_death?
=> true
```

**Comment by bgantick** [comment]
> Answered my own question:
> 
> ```ruby
> > v = Veteran.new
> > v.will_save_change_to_date_of_death?
> => false
> > v.date_of_death = Time.zone.yesterday
> => Sun, 31 Jan 2021
> > v.will_save_change_to_date_of_death?
> => true
> ```

Sorry @nanotone,  just getting around to this. Is the concern that `date_of_death_reported_at` wouldn't get populated if a method that creates a new veteran is called?

**Review by meckila-navapbc — APPROVED** [review]
LGTM

**Comment by nanotone** [comment]
@bgantick 
> Sorry @nanotone, just getting around to this. Is the concern that `date_of_death_reported_at` wouldn't get populated if a method that creates a new veteran is called?

Yep, because the initial insert of a new record might not be interpreted as a "change".

But I verified in Rails console that, for a brand new veteran, `will_save_change_to_date_of_death?` essentially treats the previous value as nil, and as long as the date_of_death gets any non-nil value, it'll return true. So we're good.

**Comment by bgantick** [comment]
> @bgantick
> 
> > Sorry @nanotone, just getting around to this. Is the concern that `date_of_death_reported_at` wouldn't get populated if a method that creates a new veteran is called?
> 
> Yep, because the initial insert of a new record might not be interpreted as a "change".
> 
> But I verified in Rails console that, for a brand new veteran, `will_save_change_to_date_of_death?` essentially treats the previous value as nil, and as long as the date_of_death gets any non-nil value, it'll return true. So we're good.

Thanks for the clarification!

**Pull request merged**